cmd/osv-scanner/main.go

```diff
@@ -121,6 +121,14 @@ func run(args []string, stdout, stderr io.Writer) int {
 				Usage:  "sets the path that local databases should be stored",
 				Hidden: true,
 			},
+			&cli.BoolFlag{
+				Name:  "experimental-all-packages",
+				Usage: "when json output is selected, prints all packages",
+			},
+			&cli.BoolFlag{
```

Comment:
I think in the json output when the licenses is specified without an allowlist, we want it to print all packages (and their licenses) by default. If that's the case, then I'm not sure how the all-packages and licenses flags should interact.

Reply:
Now that we do have a all-packages flag, maybe licenses (without allowlist) shouldn't print all packages by default, and just fill in the field for packages that already exists? If someone wants full license info for everything, pass the all-packages flag. 

@oliverchang thoughts?

Reply:
Just to be sure I understand the alternatives here, is it:

1. `--licenses` (without `--all-packages`) will just ignore all Unknown licenses?
2. `--licenses` implies `--all-packages` ?

This is complicated by the behaviour for when an allow-list is passed, because to be consistent with the vulnerability scanning behaviour, it makes more sense for `--licenses <ALLOWLIST>` to not imply `--all-packages`. 

Thinking about this a bit more, all packages should in fact have a license right? (as opposed to vulnerabilities). We just consider "Unknown" to be a license that's worthwhile reporting too, because from a license compliance perspective it's just as important. 

From that perspective, 2. above seems to make more sense to me. Not necessarily because it implies `--all-packages` for the sake of it, but rather that we do have something worthwhile to report for all packages when it comes to licenses.

Reply:
okay i see.

> This is complicated by the behaviour for when an allow-list is passed, because to be consistent with the vulnerability scanning behaviour it makes more sense for --licenses <ALLOWLIST> to not imply --all-packages.

Thinking about this, I think rather than thinking that --licenses implies --all-packages, we can think of how --licenses with an empty allowlist means every license a violation, which means all packages should be printed in the json output.

This is the solution I'm tending towards:
1. --all-packages has no effect on the table output. It will always be only the vulnerabilities in the vulns table, and if licenses is specified (regardless of allowlist), all the licenses and their counts in addition to the vulns table.
2. in json output --licenses=<allowlist> prints only packages with license violations or packages with vulns. If allowlist is empty then all packages are violating licenses (the license_violations field will be populated for every package)
3. in json output --licenses=<allowlist> with --all-packages prints all packages whether they have a vuln or license violation or not. The license_violations field is populated for every package that has a violation, license field populated for every package.

thoughts?

Reply:
The JSON output makes sense to me.

Can you clarify what happens in the table output when `--licenses` is passed in with an allowlist? 

Would the packages that are violating the allowlist be shown individually? or only a list of violating licenses in a summary format?

Reply:
I was imagining that it would be the same as if there was no allowlist specified, it would just show all the licenses with their counts. If we did want to do something, we could add a column in the table when an allowlist is specified like 'violation' with value either true or false.

Reply:
and to clarify, all licenses, including the UNKNOWN license would be in that table whether allowlist is specified or not.

Reply:
How would a user know what package violated the allowlist without having to parse the JSON?

Reply:
they wouldn't.

my thinking was listing all individual licenses, or all license violations here may clutter the view. But you do make a good point as the vulns table shows the individual vulns, but this table would only shows a summary. it doesn't mirror the behaviour of the vulns perfectly.

we did discuss in the meeting adding a summary/all-packages flag to toggle this, but iirc we concluded that we will wait for feedback first and add flags or change the table format in response to user feedback.

did you have a solution in mind?

Reply:
sorry! my bad my memory is wrong. For the table view we agreed to list a summary when allowlist is empty, and violations when allowlist isn't empty.